near-sdk-macros/src/lib.rs

```diff
@@ -316,3 +359,93 @@ pub fn function_error(item: TokenStream) -> TokenStream {
         }
     })
 }
+
+/// Adding the `events` argument for the `#[near_bindgen]` macro injects this derive macro and should be the used instead of this
+///
+/// This derive macro is used to inject the necessary wrapper and logic to auto format
+/// standard event logs. The other appropriate attribute macros are not injected with this macro.
+/// Required attributes below:
+/// ```ignore
+/// #[derive(near_sdk::serde::Serialize, std::clone::Clone)]
+/// #[serde(crate="near_sdk::serde")]
+/// #[serde(tag = "event", content = "data")]
+/// #[serde(rename_all="snake_case")]
+/// pub enum MyEvent {
+///     Event
+/// }
+/// ```
+#[proc_macro_derive(EventMetadata, attributes(event_version))]
+pub fn derive_event_attributes(item: TokenStream) -> TokenStream {
+    if let Ok(input) = syn::parse::<ItemEnum>(item) {
+        let name = &input.ident;
+        let standard_name = format!("{}_event_standard", name);
+        let standard_ident = syn::Ident::new(&standard_name, Span::call_site());
+        // version from each attribute macro
+        let mut attr_error: u8 = 0;
+        let mut event_meta: Vec<proc_macro2::TokenStream> = vec![];
+        let _ = &input.variants.iter().for_each(|var| {
+            if let Some(version) = core_impl::get_event_version(var) {
+                let var_ident = &var.ident;
+                event_meta.push(quote! {
+                    #name::#var_ident { .. } => {(#standard_ident.to_string(), #version.to_string())}
+                })
+            } else {
+                attr_error += 1;
+            }
+        });
+
+        // handle lifetimes, generics, and where clauses
+        let (impl_generics, type_generics, where_clause) = &input.generics.split_for_impl();
+        // add 'near_event lifetime for user defined events
+        let mut generics = input.generics.clone();
+        let event_lifetime = syn::Lifetime::new("'near_event", Span::call_site());
+        generics.params.insert(
+            generics.params.len(),
+            syn::GenericParam::Lifetime(syn::LifetimeDef::new(event_lifetime.clone())),
+        );
+        let (custom_impl_generics, ..) = generics.split_for_impl();
+
+        if attr_error > 0 {
+            return TokenStream::from(
+                syn::Error::new(
+                    Span::call_site(),
+                    "Near events must have `event_version`. Must have a single string literal value.",
+                )
+                .to_compile_error(),
+            );
+        }
+
+        TokenStream::from(quote! {
+
+            #[derive(near_sdk::serde::Serialize)]
+            #[serde(crate="near_sdk::serde")]
+            #[serde(rename_all="snake_case")]
+            struct EventBuilder #custom_impl_generics #where_clause {
+                standard: String,
+                version: String,
+                #[serde(flatten)]
+                event_data: &#event_lifetime #name #type_generics
+            }
```

Comment:
This will conflict if you have two event derives in the same module. Since this is a hidden type, ideally it's scoped only for the implementation of `EventJson`. Am I missing something? Why else would it be used otherwise?

Reply:
Yes, it's only used for the implementation of `EventJson` to wrap the user defined events enum and add the `standard` and `version` field with correct formatting.